cpp/src/join/conditional_join.hpp

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "join_common_utils.hpp"
+
+#include <cudf/ast/nodes.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <optional>
+
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief Computes the join operation between two tables and returns the
+ * output indices of left and right table as a combined table
+ *
+ * @param left  Table of left columns to join
+ * @param right Table of right  columns to join
+ * tables have been flipped, meaning the output indices should also be flipped
+ * @param JoinKind The type of join to be performed
+ * @param compare_nulls Controls whether null join-key values should match or not.
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ *
+ * @return Join output indices vector pair
+ */
+std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+          std::unique_ptr<rmm::device_uvector<size_type>>>
+conditional_join(table_view const& left,
+                 table_view const& right,
+                 ast::expression binary_predicate,
+                 null_equality compare_nulls,
+                 join_kind JoinKind,
```

Comment:
Nitpick: Re: `JoinKind`: Was the departure from naming conventions intentional?

Reply:
Ah that's a good call, I didn't question this at all. The JoinKind was originally a template parameter and therefore used PascalCase, and I never switched to snake_case. I can do so now.

Reply:
Just a nitpick. I was tempted to ask why the `JoinKind` isn't a template parameter. I don't know if the benefits of `constexpr` might be worth the code explosion cost. 

Thanks for clarifying.

Reply:
Yeah doubling compilation time here was a lot worse than the small cost for a runtime conditional in this case.